pkg/kv/cosmosdb/store.go

```diff
@@ -0,0 +1,428 @@
+		}
+
+		// hook for using emulator for testing
+		if params.Client == nil {
+			params.Client = http.DefaultClient
+		}
+		// Create a CosmosDB client
+		client, err = azcosmos.NewClientWithKey(params.Endpoint, cred, &azcosmos.ClientOptions{
+			ClientOptions: azcore.ClientOptions{
+				Transport: params.Client,
+			},
+		})
+		if err != nil {
+			return nil, fmt.Errorf("creating client using access key: %w", err)
+		}
+	} else {
+		cred, err := azidentity.NewDefaultAzureCredential(nil)
+		if err != nil {
+			return nil, fmt.Errorf("default creds: %w", err)
+		}
+		client, err = azcosmos.NewClient(params.Endpoint, cred, nil)
+		if err != nil {
+			return nil, fmt.Errorf("creating client with default creds: %w", err)
+		}
+	}
+
+	dbClient, err := getOrCreateDatabase(ctx, client, params)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create container client
+	containerClient, err := getOrCreateContainer(ctx, dbClient, params)
+	if err != nil {
+		return nil, err
+	}
+
+	cLevel := azcosmos.ConsistencyLevelBoundedStaleness
+	if !params.StrongConsistency {
+		cLevel = azcosmos.ConsistencyLevelSession
+	}
+	return &Store{
+		containerClient:  containerClient,
+		consistencyLevel: cLevel,
+	}, nil
+}
+
+func getOrCreateDatabase(ctx context.Context, client *azcosmos.Client, params *kvparams.CosmosDB) (*azcosmos.DatabaseClient, error) {
+	dbResp, err := client.CreateDatabase(ctx, azcosmos.DatabaseProperties{ID: params.Database}, nil)
+	if err != nil {
+		errCode := errStatusCode(err)
+		if errCode != http.StatusConflict && errCode != http.StatusCreated {
+			errPrefix := "creating database"
+			if dbResp.RawResponse != nil {
+				errPrefix += fmt.Sprintf("(%d)", dbResp.RawResponse.StatusCode)
+			}
+			return nil, fmt.Errorf("%s: %w", errPrefix, err)
+		}
+	}
+	dbClient, err := client.NewDatabase(params.Database)
+	if err != nil {
+		return nil, fmt.Errorf("creating database client: %w", err)
+	}
+	return dbClient, nil
+}
+
+func getOrCreateContainer(ctx context.Context, dbClient *azcosmos.DatabaseClient, params *kvparams.CosmosDB) (*azcosmos.ContainerClient, error) {
+	cResp, err := dbClient.CreateContainer(ctx,
+		azcosmos.ContainerProperties{
+			ID: params.Container,
+			PartitionKeyDefinition: azcosmos.PartitionKeyDefinition{
+				Paths: []string{"/partitionKey"},
+			},
+			// Excluding the value field from indexing since it is not used in queries and saves RUs for writes.
+			// partitionKey is automatically not indexed. The rest of the fields are indexed by default, including id
+			// which is unnecessary, but cannot be excluded.
+			IndexingPolicy: &azcosmos.IndexingPolicy{
+				Automatic:     false,
+				IndexingMode:  azcosmos.IndexingModeConsistent,
+				IncludedPaths: []azcosmos.IncludedPath{{Path: "/*"}},
+				ExcludedPaths: []azcosmos.ExcludedPath{{Path: "/value/?"}},
+			},
+		}, nil)
+	if err != nil {
+		errCode := errStatusCode(err)
+		if errCode != http.StatusConflict && errCode != http.StatusCreated {
```

Comment:
Same as the above comments:
- Why do we check status created on error?
- Can we leverage cosmos error to render ours? is it because their error message is too long?

Reply:
Done.